File: medium/go/1456-maximum-number-of-vowels-in-substring-of-given-length.go
package main

func maxVowels(s string, k int) int {
	vowels := map[byte]bool{
		'a':true,
		'e':true,
		'i':true,
		'o':true,
		'u':true,
	}

	currentVowels := 0 

	for i := 0; i < k; i++ {
		_, has := vowels[s[i]]

		if has {
			currentVowels++
		}
	}

	max := currentVowels

	for i:= k; i < len(s) ; i++ {
		if _, has := vowels[s[i]]; has {
			currentVowels++
		}

		if _, has := vowels[s[i-k]]; has {
			currentVowels--
		}

		if currentVowels > max {
			max = currentVowels
		}
	}

	return max 
}

File: medium/javascript/1456-maximum-number-of-vowels-in-substring-of-given-length.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxVowels = function (s, k) {
  let currentVowels = 0;
  const vowels = new Set(["a", "e", "i", "o", "u"]);

  for (let idx = 0; idx < k; idx++) {
    if (vowels.has(s[idx])) {
      currentVowels++;
    }
  }

  max = currentVowels;

  for (let idx = k; idx < s.length; idx++) {
    if (vowels.has(s[idx])) {
      currentVowels++;
    }

    if (vowels.has(s[idx - k])) {
      currentVowels--;
    }

    if (currentVowels > max) {
      max = currentVowels;
    }
  }

  return max;
};
